boot-qemu.py

```diff
@@ -772,14 +833,31 @@ def launch_qemu(cfg):
             qemu_cmd = timeout_cmd + stdbuf_cmd + qemu_cmd
 
         pretty_print_qemu_cmd(qemu_cmd)
-        try:
-            subprocess.run(qemu_cmd, check=True)
-        except subprocess.CalledProcessError as ex:
-            if ex.returncode == 124:
-                utils.red("ERROR: QEMU timed out!")
-            else:
-                utils.red("ERROR: QEMU did not exit cleanly!")
-            sys.exit(ex.returncode)
+        null_cm = contextlib.nullcontext()
+        with open(virtiofsd_log, 'w', encoding='utf-8') if share_folder_with_guest else null_cm as vfsd_log, \
+             subprocess.Popen(virtiofsd_cmd, stderr=vfsd_log, stdout=vfsd_log) if share_folder_with_guest else null_cm as vfsd_process:
+            try:
+                subprocess.run(qemu_cmd, check=True)
+            except subprocess.CalledProcessError as ex:
+                if ex.returncode == 124:
+                    utils.red("ERROR: QEMU timed out!")
+                else:
+                    utils.red("ERROR: QEMU did not exit cleanly!")
+                    # If virtiofsd is dead, it is pretty likely that it was the
+                    # cause of QEMU failing so add to the existing exception using
+                    # 'from'.
+                    if vfsd_process and vfsd_process.poll():
+                        vfsd_log_txt = virtiofsd_log.read_text(
+                            encoding='utf-8')
+                        raise Exception(
+                            f"virtiofsd failed with: {vfsd_log_txt}") from ex
+                sys.exit(ex.returncode)
+            finally:
+                if vfsd_process:
+                    vfsd_process.kill()
+                    # Delete the memory to save space, it does not have to be
+                    # persistent
```

Comment:
The indentation here is getting pretty deep. Please move this into a new function. Can be a patch on top.

Reply:
I can look at refactoring running QEMU for `gdb` vs. regularly into separate functions?

Reply:
2eca31e3da3c492cf71df4bfb3099eda7b056f10